### PR TITLE
CP-26856: unbind pciback driver before disable SR-IOV

### PR DESCRIPTION
**Pull request by xiewei20082008**
Signed-off-by: Wei Xie <wei.xie@citrix.com>

---
This change is [Reviewable](https://reviewable.io/reviews/xapi-project/xcp-networkd/129)

### CODE REVIEW

File: lib/network_utils.ml [hunk 1]
```diff
@@ -246,21 +250,61 @@ module Sysfs = struct
 			Result.Ok devices.(0)
 		with _ -> Result.Error (Parent_device_of_vf_not_found, "Can not get parent device for " ^ pcibuspath)
 
-	let device_index_of_vf parent_device pcibuspath =
+	let get_child_vfs_sysfs_paths dev =
```
Comment (xiewei20082008) [inline lib/network_utils.ml]: As both functions `device_index_of_vf` and `unbind_child_vfs` need to iterate the VF sysfs paths, I abstract this function `get_child_vfs_sysfs_paths` out from original `device_index_of_vf`

**Comment by xiewei20082008** [comment]
Review status: 0 of 3 files reviewed at latest revision, 5 unresolved discussions.

---

*[lib/network_utils.ml, line 289 at r1](https://reviewable.io:443/reviews/xapi-project/xcp-networkd/129#-L5DYatTPoqy5Ompsuae-r1-289:-L5GzaSZZmD_uNw1vTwL:b-dttw8v) ([raw file](https://github.com/xapi-project/xcp-networkd/blob/0473acf0b829d5889525735db9b8948ca1be87b8/lib/network_utils.ml#L289)):*
<details><summary><i>Previously, robhoes (Rob Hoes) wrote…</i></summary><blockquote>

Why don't we just want to unbind any driver? The goal here is to switch of SR-IOV completely, so I think we can use a big hammer.
</blockquote></details>

I was thinking about a case where users manually use VFs in Dom0 as a NIC, then a VF can be bound to a VF driver like `igbvf`. For that case, I think it is better to notify users `VF %s is bound to an unexpected driver` rather than to make a user find his VF NIC just disappeared.

---

*[lib/network_utils.ml, line 306 at r1](https://reviewable.io:443/reviews/xapi-project/xcp-networkd/129#-L5DYatTPoqy5Ompsuae-r1-306:-L5H3d1Rq4o2j0J2oiqo:b10un11) ([raw file](https://github.com/xapi-project/xcp-networkd/blob/0473acf0b829d5889525735db9b8948ca1be87b8/lib/network_utils.ml#L306)):*
<details><summary><i>Previously, robhoes (Rob Hoes) wrote…</i></summary><blockquote>

Nice, you're getting quite good with the `>>=` :)
</blockquote></details>

Haha, 'a -> 'b -> 'a is really good pattern. :)

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/xapi-project/xcp-networkd/129)*

**Comment by robhoes** [comment]
Reviewed 3 of 3 files at r1.
Review status: all files reviewed at latest revision, 4 unresolved discussions.

---

*[lib/network_utils.ml, line 289 at r1](https://reviewable.io:443/reviews/xapi-project/xcp-networkd/129#-L5DYatTPoqy5Ompsuae-r1-289:-L5ImlaaxuwRjUBo7iYe:b-w8h5q7) ([raw file](https://github.com/xapi-project/xcp-networkd/blob/0473acf0b829d5889525735db9b8948ca1be87b8/lib/network_utils.ml#L289)):*
<details><summary><i>Previously, xiewei20082008 (Wei Xie) wrote…</i></summary><blockquote>

I was thinking about a case where users manually use VFs in Dom0 as a NIC, then a VF can be bound to a VF driver like `igbvf`. For that case, I think it is better to notify users `VF %s is bound to an unexpected driver` rather than to make a user find his VF NIC just disappeared.
</blockquote></details>

I don't think we can support such use cases. Users have to make a choice: either let xapi/networkd manage the NIC, or not, but not half/half. It would create all sorts of complications.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/xapi-project/xcp-networkd/129)*

File: lib/network_utils.ml [hunk 2]
```diff
@@ -246,21 +250,61 @@ module Sysfs = struct
...
+				|> List.map (Filename.concat device_path)
+			)
+		with _ -> Result.Error (Vf_sysfs_path_not_found, "Can not get child vfs sysfs paths for " ^ dev)
+
+	let device_index_of_vf parent_dev pcibuspath =
+		try
+			let open Rresult.R.Infix in
+			get_child_vfs_sysfs_paths parent_dev >>= fun paths ->
+			let group = 
+				List.find (fun x -> Xstringext.String.has_substr (Unix.readlink x) pcibuspath) paths
```
Comment (mseri) [inline lib/network_utils.ml]: `xcp-networkd` has been ported away from `Xstringext`.
Please do not use `Xstringext` and instead use 
```ocaml
Atring.String.is_infix ~affix:(Unix.readlink x)
```

**Comment by xiewei20082008** [comment]
Review status: 1 of 3 files reviewed at latest revision, 5 unresolved discussions.

---

*[lib/network_utils.ml, line 269 at r1](https://reviewable.io:443/reviews/xapi-project/xcp-networkd/129#-L5DYatTPoqy5Ompsuae-r1-269:-L5wkqYrLNCHyKgGvYpT:b-ioxi1u) ([raw file](https://github.com/xapi-project/xcp-networkd/blob/0473acf0b829d5889525735db9b8948ca1be87b8/lib/network_utils.ml#L269)):*
<details><summary><i>Previously, mseri (Marcello Seri) wrote…</i></summary><blockquote>

`xcp-networkd` has been ported away from `Xstringext`.
Please do not use `Xstringext` and instead use 
```ocaml
Atring.String.is_infix ~affix:(Unix.readlink x)
```
</blockquote></details>

Done. Thank you.

---

*[lib/network_utils.ml, line 289 at r1](https://reviewable.io:443/reviews/xapi-project/xcp-networkd/129#-L5DYatTPoqy5Ompsuae-r1-289:-L5wkzOxlNlGGPiNGRAj:bgkgn4a) ([raw file](https://github.com/xapi-project/xcp-networkd/blob/0473acf0b829d5889525735db9b8948ca1be87b8/lib/network_utils.ml#L289)):*
<details><summary><i>Previously, robhoes (Rob Hoes) wrote…</i></summary><blockquote>

I don't think we can support such use cases. Users have to make a choice: either let xapi/networkd manage the NIC, or not, but not half/half. It would create all sorts of complications.
</blockquote></details>

Done. Understood.

---

*[lib/network_utils.ml, line 294 at r1](https://reviewable.io:443/reviews/xapi-project/xcp-networkd/129#-L5DYatTPoqy5Ompsuae-r1-294:-L5wl7wRc-A1j8MxVB2h:b-896fix) ([raw file](https://github.com/xapi-project/xcp-networkd/blob/0473acf0b829d5889525735db9b8948ca1be87b8/lib/network_utils.ml#L294)):*
<details><summary><i>Previously, robhoes (Rob Hoes) wrote…</i></summary><blockquote>

Instead of using the fixed `!pciback_unbind_interface`, can't we just use something like `vf_path ^ "/driver/unbind"`, which will work for pciback , and other drivers as well (usually)? I'm not sure about `remove_slot`, but we can do that as best-effort.
</blockquote></details>

Done.

---

*[networkd/networkd.ml, line 64 at r1](https://reviewable.io:443/reviews/xapi-project/xcp-networkd/129#-L5DYatTPoqy5Ompsuag-r1-64:-L5wlW5a8QnKJQmnrd4Q:boyymvz) ([raw file](https://github.com/xapi-project/xcp-networkd/blob/0473acf0b829d5889525735db9b8948ca1be87b8/networkd/networkd.ml#L64)):*
<details><summary><i>Previously, robhoes (Rob Hoes) wrote…</i></summary><blockquote>

Hmm... separate commits for additional fixes please. I had to stare at this for a while to see what had changed or whether it was just whitespace.
</blockquote></details>

Done. Splitted it into two commits.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/xapi-project/xcp-networkd/129)*

**Comment by robhoes** [comment]
Reviewed 2 of 2 files at r2.
Review status: all files reviewed at latest revision, 6 unresolved discussions.

---

*[lib/network_utils.ml, line 269 at r1](https://reviewable.io:443/reviews/xapi-project/xcp-networkd/129#-L5DYatTPoqy5Ompsuae-r1-269:-L5yTpkLqoYtAZAzBEEG:bwetr9r) ([raw file](https://github.com/xapi-project/xcp-networkd/blob/0473acf0b829d5889525735db9b8948ca1be87b8/lib/network_utils.ml#L269)):*
<details><summary><i>Previously, xiewei20082008 (Wei Xie) wrote…</i></summary><blockquote>

Done. Thank you.
</blockquote></details>

Good, looks like you've correctly Marcello's version as well (he said).

---

*[lib/network_utils.ml, line 290 at r2](https://reviewable.io:443/reviews/xapi-project/xcp-networkd/129#-L5yUR04ZIqAK1gBDdKv:-L5yUR04ZIqAK1gBDdKw:b-14j6ms) ([raw file](https://github.com/xapi-project/xcp-networkd/blob/6c53ae845d4a321956d83f6bfdfd9d5ef13085f9/lib/network_utils.ml#L290)):*
> ```OCaml
> 			(* not bind to any driver, Ok *)
> 			| "" -> Result.Ok ()
> 			| driver -> 
> ```

This is a little confusing now, because now you have `driver` and `driver_name` for the same thing. It's better with an if-statement (`if driver_name = "" then .. else ..`).

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/xapi-project/xcp-networkd/129)*

**Comment by xiewei20082008** [comment]
Review status: 2 of 3 files reviewed at latest revision, 3 unresolved discussions.

---

*[lib/network_utils.ml, line 290 at r2](https://reviewable.io:443/reviews/xapi-project/xcp-networkd/129#-L5yUR04ZIqAK1gBDdKv:-L6-RkQOywybSIhUmRSs:b-896fix) ([raw file](https://github.com/xapi-project/xcp-networkd/blob/6c53ae845d4a321956d83f6bfdfd9d5ef13085f9/lib/network_utils.ml#L290)):*
<details><summary><i>Previously, robhoes (Rob Hoes) wrote…</i></summary><blockquote>

This is a little confusing now, because now you have `driver` and `driver_name` for the same thing. It's better with an if-statement (`if driver_name = "" then .. else ..`).
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/xapi-project/xcp-networkd/129)*

**Comment by robhoes** [comment]
:lgtm:

---

Reviewed 1 of 1 files at r3.
Review status: all files reviewed at latest revision, 2 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/xapi-project/xcp-networkd/129#-:-L61MbpokW1kbetbbFbt:bnfp4nl)*

**Pull request merged**